# Hypothesis: I believe students who attend office hours are less likely to recommend the course because they needed extra help with the course. 

__author__ = "730247598

 ## First I will enable autoreload to automatically reload any changes I make and save in **data_utilities.py**.

In [151]:
%reload_ext autoreload
%autoreload 2

## Then I will import in the survey data table.

In [152]:
from data_utilities import read_csv_rows
from __future__ import annotations

DATA_DIRECTORY = "../../data"
DATA_FILE_PATH = f"{DATA_DIRECTORY}/survey.csv"

data_rows: list[dict[str, str]] = read_csv_rows('../../data/survey.csv')

if len(data_rows) == 0:
    print("Go implement read_csv_rows in data_utils.py")
else:
    print(f"Data File Read: {DATA_FILE_PATH}")
    print(f"{len(data_rows)} rows")
    print(f"{len(data_rows[0].keys())} columns")
    print(f"Columns names: {data_rows[0].keys()}")

Data File Read: ../../data/survey.csv
573 rows
36 columns
Columns names: dict_keys(['row_number', 'year', 'unc_status', 'comp_major', 'primary_major', 'data_science', 'prereqs', 'prior_exp', 'AP_Principles', 'AP_A', 'other_comp', 'prior_time', 'languages', 'section', 'in_person_classes', 'hours_online_social', 'hours_online_work', 'lesson_time', 'sync_perf', 'all_sync', 'own_notes', 'own_examples', 'oh_visits', 'ls_effective', 'lsqs_effective', 'programming_effective', 'qz_effective', 'oh_effective', 'tutoring_effective', 'kaki_effective', 'pace', 'difficulty', 'understanding', 'interesting', 'valuable', 'would_recommend'])


## In order to manipulate the data in the table easier, I will convert the table from a list of rows into a dictionary of columns.

In [153]:
from data_utilities import columnar

data_cols: dict[str, list[str]] = columnar(data_rows)

if len(data_cols.keys()) == 0:
    print("Error: empty table")
else:
    print(f"{len(data_cols.keys())} columns")
    print(f"{len(data_cols['row_number'])} rows")
    print(f"Columns names: {data_cols.keys()}")

36 columns
573 rows
Columns names: dict_keys(['row_number', 'year', 'unc_status', 'comp_major', 'primary_major', 'data_science', 'prereqs', 'prior_exp', 'AP_Principles', 'AP_A', 'other_comp', 'prior_time', 'languages', 'section', 'in_person_classes', 'hours_online_social', 'hours_online_work', 'lesson_time', 'sync_perf', 'all_sync', 'own_notes', 'own_examples', 'oh_visits', 'ls_effective', 'lsqs_effective', 'programming_effective', 'qz_effective', 'oh_effective', 'tutoring_effective', 'kaki_effective', 'pace', 'difficulty', 'understanding', 'interesting', 'valuable', 'would_recommend'])


## After that, I'll print the first 5 values in each of the columns to make sure I'm getting the data I want. For better readability, I'll tabulate the data.

In [154]:
from data_utilities import head
from tabulate import tabulate

data_cols_head: dict[str, list[str]] = head(data_cols, 5)

if len(data_cols_head.keys()) != len(data_cols.keys()) or len(data_cols_head["row_number"]) != 10:
    print("Complete your implementation of columnar in data_utils.py")
    print("Be sure to follow the guidelines above and save your work before re-evaluating!")


tabulate(data_cols_head, data_cols_head.keys(), "html")

Complete your implementation of columnar in data_utils.py
Be sure to follow the guidelines above and save your work before re-evaluating!


'<table>\n<thead>\n<tr><th style="text-align: right;">  row_number</th><th style="text-align: right;">  year</th><th>unc_status           </th><th>comp_major  </th><th>primary_major                   </th><th>data_science  </th><th>prereqs                                          </th><th>prior_exp                   </th><th>AP_Principles  </th><th>AP_A  </th><th>other_comp  </th><th>prior_time  </th><th>languages  </th><th>section  </th><th>in_person_classes  </th><th>hours_online_social  </th><th>hours_online_work  </th><th style="text-align: right;">  lesson_time</th><th style="text-align: right;">  sync_perf</th><th style="text-align: right;">  all_sync</th><th style="text-align: right;">  own_notes</th><th style="text-align: right;">  own_examples</th><th style="text-align: right;">  oh_visits</th><th style="text-align: right;">  ls_effective</th><th style="text-align: right;">  lsqs_effective</th><th style="text-align: right;">  programming_effective</th><th style="text-align: right;">  qz_effective</th><th>oh_effective  </th><th>tutoring_effective  </th><th>kaki_effective  </th><th style="text-align: right;">  pace</th><th style="text-align: right;">  difficulty</th><th style="text-align: right;">  understanding</th><th style="text-align: right;">  interesting</th><th style="text-align: right;">  valuable</th><th style="text-align: right;">  would_recommend</th></tr>\n</thead>\n<tbody>\n<tr><td style="text-align: right;">           0</td><td style="text-align: right;">    23</td><td>Returning UNC Student</td><td>No          </td><td>Advertising and Public Relations</td><td>No            </td><td>MATH 129P, MATH 130                              </td><td>None to less than one month!</td><td>               </td><td>      </td><td>            </td><td>            </td><td>           </td><td>In-Person</td><td>                   </td><td>3 to 5 hours         </td><td>3 to 5 hours       </td><td style="text-align: right;">            5</td><td style="text-align: right;">          2</td><td style="text-align: right;">         2</td><td style="text-align: right;">          4</td><td style="text-align: right;">             2</td><td style="text-align: right;">          2</td><td style="text-align: right;">             7</td><td style="text-align: right;">               6</td><td style="text-align: right;">                      6</td><td style="text-align: right;">             5</td><td>5             </td><td>                    </td><td>3               </td><td style="text-align: right;">     5</td><td style="text-align: right;">           5</td><td style="text-align: right;">              4</td><td style="text-align: right;">            5</td><td style="text-align: right;">         6</td><td style="text-align: right;">                7</td></tr>\n<tr><td style="text-align: right;">           1</td><td style="text-align: right;">    24</td><td>Returning UNC Student</td><td>No          </td><td>Anthropology                    </td><td>No            </td><td>MATH 129P, MATH 231                              </td><td>None to less than one month!</td><td>               </td><td>      </td><td>            </td><td>            </td><td>           </td><td>In-Person</td><td>                   </td><td>5 to 10 hours        </td><td>5 to 10 hours      </td><td style="text-align: right;">            4</td><td style="text-align: right;">          5</td><td style="text-align: right;">         4</td><td style="text-align: right;">          7</td><td style="text-align: right;">             6</td><td style="text-align: right;">          0</td><td style="text-align: right;">             6</td><td style="text-align: right;">               5</td><td style="text-align: right;">                      7</td><td style="text-align: right;">             5</td><td>              </td><td>                    </td><td>6               </td><td style="text-align: right;">     5</td><td style="text-align: right;">           3</td><td style="text-align: right;">

## Next, I will isolate just the **OH_visits** column and the **would_recommend** column.

In [155]:
from data_utilities import select

selected_data: dict[str, list[str]] = select(data_cols, ["oh_visits", "would_recommend"])

tabulate(head(selected_data, 10), selected_data.keys(), "html")

'<table>\n<thead>\n<tr><th style="text-align: right;">  oh_visits</th><th style="text-align: right;">  would_recommend</th></tr>\n</thead>\n<tbody>\n<tr><td style="text-align: right;">          2</td><td style="text-align: right;">                7</td></tr>\n<tr><td style="text-align: right;">          0</td><td style="text-align: right;">                7</td></tr>\n</tbody>\n</table>'

## I will take all the values in the oh_visits column and add them to a list called oh_visit_values.

In [156]:
from data_utilities import column_values

oh_visit_values: list[str] = column_values(data_rows, "oh_visits")

if len(oh_visit_values) == 0:
    print("Complete your implementation of column_values in data_utils.py")
    print("Be sure to follow the guidelines above and save your work before re-evaluating!")
else:
    print(f"Column 'oh_visits' has {len(oh_visit_values)} values.")
    print("The first five values are:")
    for i in range(5):
        print(oh_visit_values[i])

Column 'oh_visits' has 573 values.
The first five values are:
2
0
0
4
1


Then I will convert the oh_visits_values to a list of ints.

In [157]:
for i in range(0, len(oh_visit_values)):
    oh_visit_values[i] = int(oh_visit_values[i])

print(oh_visit_values)

[2, 0, 0, 4, 1, 1, 0, 1, 1, 2, 3, 2, 0, 3, 1, 0, 1, 2, 1, 2, 0, 0, 1, 0, 0, 0, 0, 1, 2, 0, 3, 1, 0, 2, 2, 1, 1, 0, 0, 1, 1, 2, 1, 2, 1, 1, 0, 1, 2, 2, 3, 1, 1, 1, 0, 1, 1, 1, 0, 1, 0, 1, 2, 2, 2, 1, 0, 1, 1, 4, 0, 3, 3, 5, 1, 1, 0, 0, 1, 4, 0, 2, 0, 2, 1, 1, 0, 0, 1, 0, 1, 1, 2, 2, 0, 0, 0, 0, 1, 2, 1, 1, 0, 1, 0, 1, 3, 3, 1, 2, 1, 0, 1, 1, 1, 0, 2, 1, 0, 1, 0, 5, 4, 1, 0, 3, 5, 5, 1, 2, 4, 2, 2, 0, 1, 5, 1, 0, 1, 3, 1, 4, 2, 0, 0, 2, 0, 1, 2, 0, 4, 0, 3, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 2, 2, 2, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 1, 1, 2, 0, 1, 1, 0, 1, 0, 0, 2, 0, 0, 0, 1, 0, 0, 0, 1, 1, 2, 0, 0, 0, 4, 1, 3, 0, 3, 0, 2, 0, 0, 1, 1, 0, 2, 1, 0, 1, 1, 1, 1, 0, 2, 2, 3, 5, 5, 3, 0, 0, 1, 0, 0, 1, 0, 2, 1, 0, 0, 1, 0, 2, 1, 2, 1, 1, 1, 0, 1, 1, 2, 0, 2, 0, 3, 0, 3, 4, 4, 0, 1, 3, 0, 0, 1, 0, 5, 0, 1, 5, 1, 1, 2, 2, 5, 5, 0, 3, 0, 0, 2, 2, 5, 2, 0, 0, 2, 1, 2, 0, 0, 1, 1, 2, 2, 2, 2, 2, 1, 0, 0, 4, 2, 1, 2, 1, 1, 1, 0, 1, 2, 2, 2, 1, 1, 0, 4, 1, 1, 0, 1, 1, 2, 2, 1, 0, 3, 0, 2, 0, 0, 0, 1, 

I will find out what **would_recommend** values each student that goes to office hours at least once per assignment inputed and sum them.

In [158]:
oh_students_would_rec_score: int = 0

for i in range(0, len(oh_visit_values)):
    if oh_visit_values[i] > 0:
        oh_students_would_rec_score += would_rec_values[i]

print(oh_students_would_rec_score)

2012


Then I will find the number of people in the class who visit office hours at least one time per assignment.

In [159]:
oh_vis_students: int = 0

for i in range(0, len(oh_visit_values)):
    if oh_visit_values[i] > 0:
        oh_vis_students += 1

print(oh_vis_students)

388


Lastly I will find the average would_recommend score for the students who visited office hours by dividing the would recommend score from the office hour students by the number of students that visit office hours.

In [160]:
oh_vis_avg: float = oh_students_would_rec_score / oh_vis_students

print(oh_vis_avg)

5.185567010309279


## I will take all of the values in the would_recommend column and put them in a list.

In [161]:
would_rec_values: list[str] = column_values(data_rows, "would_recommend")

if len(would_rec_values) == 0:
    print("Complete your implementation of column_values in data_utils.py")
    print("Be sure to follow the guidelines above and save your work before re-evaluating!")
else:
    print(f"Column 'would_recommend' has {len(would_rec_values)} values.")
    print("The first five values are:")
    for i in range(5):
        print(would_rec_values[i])

Column 'would_recommend' has 573 values.
The first five values are:
7
7
7
7
7


Then I will convert would_rec_values to a list of ints.

In [162]:
for i in range(0, len(would_rec_values)):
    would_rec_values[i] = int(would_rec_values[i])

print(would_rec_values)


[7, 7, 7, 7, 7, 4, 6, 4, 6, 5, 4, 4, 5, 6, 6, 6, 7, 5, 1, 3, 7, 7, 7, 7, 5, 5, 6, 4, 5, 6, 5, 5, 7, 5, 2, 3, 4, 7, 4, 2, 5, 3, 4, 7, 7, 5, 7, 6, 6, 5, 2, 6, 2, 3, 5, 7, 5, 7, 5, 5, 6, 6, 7, 6, 4, 5, 7, 7, 5, 5, 6, 5, 4, 4, 7, 7, 7, 6, 6, 7, 6, 4, 7, 4, 7, 4, 5, 3, 6, 6, 5, 3, 3, 6, 7, 2, 7, 7, 6, 3, 5, 7, 7, 3, 7, 3, 6, 7, 4, 7, 7, 7, 5, 5, 7, 5, 6, 5, 7, 5, 7, 5, 3, 7, 7, 4, 1, 5, 4, 4, 4, 7, 7, 7, 7, 7, 6, 6, 7, 7, 7, 6, 5, 7, 7, 4, 7, 7, 5, 7, 6, 5, 6, 7, 6, 7, 7, 7, 7, 7, 7, 6, 7, 7, 7, 7, 6, 6, 7, 7, 5, 7, 7, 5, 6, 6, 6, 6, 7, 7, 7, 7, 5, 5, 6, 7, 6, 6, 7, 7, 5, 7, 7, 7, 2, 7, 7, 6, 5, 7, 7, 7, 7, 7, 6, 5, 6, 6, 7, 7, 6, 1, 7, 7, 7, 5, 6, 7, 7, 3, 7, 7, 6, 3, 6, 4, 6, 6, 7, 6, 4, 6, 5, 7, 7, 7, 7, 6, 6, 6, 6, 6, 6, 7, 6, 7, 7, 7, 5, 7, 7, 6, 3, 7, 4, 7, 6, 5, 7, 6, 7, 7, 7, 7, 5, 4, 6, 7, 4, 6, 5, 7, 4, 6, 7, 7, 6, 1, 7, 6, 5, 5, 6, 6, 4, 7, 6, 1, 5, 6, 4, 2, 4, 7, 2, 7, 5, 6, 4, 7, 5, 5, 4, 6, 4, 4, 4, 5, 6, 7, 4, 6, 3, 5, 5, 6, 3, 7, 5, 7, 7, 6, 6, 1, 6, 7, 6, 7, 7, 7, 6, 7, 7, 

Finally I will calculate sum the total recommendation score of the students and then divide that score by the number of students.

In [163]:
would_rec_sum: int = sum(would_rec_values)

would_rec_avg: float = would_rec_sum / len(would_rec_values)

print(would_rec_sum)
print(would_rec_avg)

3141
5.481675392670157


# Summary

Because the students who visit office hours scored a 5.18 on the 'would recommend this course' scale compared to the class average of 5.48, I believe my hypothesis was correct.